Core/clim-core/gadgets.lisp

```diff
@@ -1659,8 +1662,19 @@ and must never be nil.")
 
 (defmethod compose-space ((pane slider-pane) &key width height)
   (declare (ignore width height))
-  (let ((minor (+ 8 (* 2 4) (if (gadget-show-value-p pane) 30 0)))
-        (major 128))
+  (let* ((vw
+           (max ; no slider value can be wider than widest of min-value or max-value
+             (text-size pane
+                        (format-value (gadget-min-value pane) (slider-decimal-places pane)))
+             (text-size pane
+                        (format-value (gadget-max-value pane) (slider-decimal-places pane)))))
+         (vh
+           (text-style-ascent (pane-text-style pane) pane))
+         (minor
+           (if (gadget-show-value-p pane)
+               (* 2 (+ 10.0 (if (eq (gadget-orientation pane) :vertical) vw vh))) ; the value
+               (* 2 (1+ 8.0)))) ; the knob
+         (major 128))
```

Comment:
I think that this change (in general) operates on top of invalid assumption that min/max value is the narrowest/widest (especially if we agree, that text-style may be something else than `:fix`).

For instance, let's take range from 0 to 111 - in this case 100 is wider than both (assuming `0` is wider than `1`).
As of vertical panes ascent suggests, that font will never have numbers having pixels /below/ the baseline. I think it is a reasonable assumption, but point still stands.

I think that instead of having vw and vh values this should be rewritten to take into account the worst case scenario using provided clim primitives for that. I think that something in this spirit should be OK:

```
(let* ((value-size (etypecase (gadget-orientation pane)
                     (:vertical (text-style-height (pane-text-style pane) pane))
                     (:horizontal (let* ((dp (slider-decimal-places pane))
                                         (s1 (format-value (gadget-min-value pane) dp))
                                         (s2 (format-value (gadget-max-value pane) dp)))
                                    (* (max (length s1) (length s2))
                                       (text-style-width (pane-text-style pane) pane))))))
       (minor (if (gadget-show-value-p pane)
                  (* 2 (+ 10.0 value-size))
                  (* 2 (1+ 8.0))))
       (major 128))
...)
```

note that in this case compose-space will demand more space, because height is factually `(+ ascent descent)` and width is `(* number-of-characters size-of-letter-m)`. But IMHO this approach is more general than probing for text-size of min/max values.

Reply:
also I'd add better comments where these numeric constants come from. 10.0, 8.0, why 1+? why *2? all this is not obvious. I'll look into main part with text-styles for merging tomorrow.

Reply:
Thanks for your thoughtful suggestions. In my testing I hadn't noticed intermediate values being wider than min/max because DejaVu Sans (and it turns out many other fonts) have decimal digits that are all the same width. However, I believe the following deals correctly with variable width digits. It probes each digit individually since in some fonts even the widest digit is significantly narrower than the letter "m". For the other dimension, I used the font ascent not the height because draw-text* in handle-repaint draws with y alignment `:baseline`, not `:bottom`.
```
(defmethod compose-space ((pane slider-pane) &key width height)
  (declare (ignore width height))
  (let* ((value-size (ecase (gadget-orientation pane)
                       (:horizontal (text-style-ascent (pane-text-style pane) pane))
                       (:vertical (let* ((dp (slider-decimal-places pane))
                                         (s1 (format-value (gadget-min-value pane) dp))
                                         (s2 (format-value (gadget-max-value pane) dp)))
                                    (* (max (length s1) (length s2))
                                       (loop for c across "0123456789.-"
                                             maximize (text-size pane c)))))))
          (minor (if (gadget-show-value-p pane)
                     (* 2 (+ 10.0 value-size)) ; the value
                     (* 2 (1+ 8.0)))) ; the knob
          (major 128))
    (if (eq (gadget-orientation pane) :vertical)
        (make-space-requirement :min-width  minor :width  minor
                                :min-height major :height major)
        (make-space-requirement :min-width  major :width  major
                                :min-height minor :height minor))))
```
The 10.0 and 8.0 are used in the same way as in handle-repaint. The 1+ corresponds to the draw-circle* call that includes 1+ and 8.0. The 2 is because sliders are drawn with the same amount of space on one side of the line as the other. These values are more obvious when looked at in the context of the complete source code file.

Reply:
On further reflection, a McCLIM programmer would be ill-advised to use sliders with `:show-value-p t` in conjunction with a proportional font that has non-monospaced digits. If they do, they will get the unpleasant visual effect described in the article [Proportional vs. Monospaced Numbers: When to use which one in order to avoid "Wiggling Labels"](https://blog.usejournal.com/proportional-vs-monospaced-numbers-when-to-use-which-one-in-order-to-avoid-wiggling-labels-e31b1c83e4d0)

Reply:
thanks for the interesting link!  regardless, your approach with maximize and text-size looks good to me. ill advised or not, this is not our problem.

regarding magic values: my intention was to suggest, that you should put comment describing where they come from (i.e `magic values taken from handler-repaint`).

Reply:
OK, so the problem originated in extensions/fonts module. Please see branch default-text-style-fixes for a commit which fixes issue #624. It also modifies text-style initarg to be put directly in basic-gadget (only once) -- second commit. Please rebase your remaining changes (`compose-space` fixes and `suggested-height` against that branch).

I'm not doing that myself because that would be kind of hijack of the pull request.